proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+-->
+
+## Introduction
+
+In this document, we propose adding a new function called `move` to the swift
+standard library, which ends the lifetime of a specific local `let`,
+local `var`, or `consuming` parameter. In order to enforce this, the compiler will
+emit a flow-sensitive diagnostic upon any uses that are after the move
+function. As an example:
+
+```
+useX(x) // do some stuff with local variable x
+
+// Ends lifetime of x, y's lifetime begins.
+let y = move(x) // [1]
+
+useY(y) // do some stuff with local variable y
+useX(x) // error, x's lifetime was ended at [1]
+
+// Ends lifetime of y. Since _ is no-op, we perform an actual release here.
+let _ = move(y) // [2]
+useX(x) // error, x's lifetime was ended at [1]
+useY(y) // error, y's lifetime was ended at [2]
+```
+
+This allows the user to influence the uniqueness of COW data structures
+and reduce retain/release calls in a way that is future-proof against changes
+to the language implementation or changes to the source code that unexpectedly
+extend the lifetime of moved variables. Consider the following array/uniqueness
+example:
+
+```
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  consumeFinalY(y)
+
+  // We no longer use y after this point. Ideally, x would be guaranteed
+  // unique so we know we can append again without copying.
+  x.append(7)
+}
+```
+
+in the example above, without `move`, `y`'s formal lifetime extends to the end of
+scope. When we go back to using `x`, although the compiler may optimize
+the actual lifetime of `y` to release it after its last use, there isn't
+a strong guarantee that it will. Even if the optimizer does what we want,
+programmers modifying this code in the future
+may introduce new references to `y` that inadvertently extend its lifetime
+and break our attempt to keep `x` unique.
+
+Swift-evolution pitch thread: [https://forums.swift.org/t/pitch-move-function-use-after-move-diagnostic](https://forums.swift.org/t/pitch-move-function-use-after-move-diagnostic)
+
+
+## Proposed solution: Move Function + "Use After Move" Diagnostic
+
+That is where the `move` function comes into play. The `move` function is a new
+generic stdlib function that when given a local let, local var, or parameter
+argument provides a compiler guarantee to the programmer that the binding will
```

Comment:
This is a little oddly phrased: "The move function provides a _compiler_ guarantee to the _programmer_?"  I would rather say that the programmer can insert `move` operations in order to explicitly end the lifetime of a particular binding.  This provides guaranteed behavior in the face of future changes to the surrounding code or to the compiler optimization pipeline.